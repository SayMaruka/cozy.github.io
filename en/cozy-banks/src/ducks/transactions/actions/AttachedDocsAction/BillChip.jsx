/* global __DEVELOPMENT__ */

import React from 'react'
import PropTypes from 'prop-types'
import { translate } from 'cozy-ui/transpiled/react'
import Chip from 'cozy-ui/transpiled/react/Chip'
import flag from 'cozy-flags'
import FileOpener from 'ducks/transactions/FileOpener'
import FileIcon from 'ducks/transactions/actions/AttachedDocsAction/FileIcon'
import { Figure } from 'components/Figure'
import { AugmentedModalOpener, isAugmentedModalBill } from 'ducks/demo'
import { getBrands } from 'ducks/brandDictionary'

export class DumbBillChip extends React.PureComponent {
  static propTypes = {
    bill: PropTypes.object.isRequired
  }

  getInvoiceId(bill) {
    if (!bill.invoice) {
      if (__DEVELOPMENT__) {
        console.warn('Bill without invoice', bill) // eslint-disable-line no-console
      }
      throw new Error('Bill without invoice')
    }

    const [doctype, id] = bill.invoice.split(':')

    if (!doctype || !id) {
      throw new Error('Invoice is malformed. invoice: ' + bill.invoice)
    }

    return [doctype, id]
  }

  getChipLabel(vendorName) {
    const { t } = this.props

    return vendorName
      ? t('Transactions.actions.attachedDocs.billWithVendor', {
          vendorName: vendorName
        })
      : t('Transactions.actions.attachedDocs.billWithoutVendor')
  }

  render() {
    const { bill, transaction } = this.props

    let invoiceId

    try {
      invoiceId = this.getInvoiceId(bill)[1]
    } catch (err) {
      // eslint-disable-next-line no-console
      console.warn(err)
      return null
    }

    const shouldUseAugmentedModal = flag('demo') && isAugmentedModalBill(bill)

    const Wrapper = shouldUseAugmentedModal ? AugmentedModalOpener : FileOpener

    // Bill's vendor can be a slug. We get the brand from our dictionary to be
    // sure that we show the brand name and not a konnector slug
    const [brand] = getBrands(
      brand => brand.name === bill.vendor || brand.konnectorSlug === bill.vendor
    )
    const vendorName = brand && brand.name

    return (
      <Wrapper fileId={invoiceId} key={invoiceId} transaction={transaction}>
        <Chip component="button" size="small" variant="outlined">
          <FileIcon
            color={bill.isRefund ? 'var(--emerald)' : undefined}
            className="u-flex-shrink-0"
          />
          {bill.isRefund ? (
            <>
              {vendorName && (
                <span className="u-valid u-mr-half">{vendorName}</span>
              )}
              <Figure total={bill.amount} coloredPositive signed symbol="â‚¬" />
            </>
          ) : (
            this.getChipLabel(vendorName)
          )}
        </Chip>
      </Wrapper>
    )
  }
}

const BillChip = translate()(DumbBillChip)

export default BillChip
