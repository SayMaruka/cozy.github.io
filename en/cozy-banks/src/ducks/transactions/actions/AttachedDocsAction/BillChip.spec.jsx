import React from 'react'
import { DumbBillChip } from './BillChip'
import { shallow } from 'enzyme'

jest.mock('ducks/transactions/FileOpener', () => ({ children }) => children)

describe('BillChip', () => {
  describe('when the bill has no invoice', () => {
    beforeEach(() => {
      jest.spyOn(console, 'warn').mockImplementation(() => {})
    })

    afterEach(() => {
      // eslint-disable-next-line no-console
      console.warn.mockRestore()
    })

    it('should render nothing', () => {
      const bill = {
        _id: 'fakebill'
      }

      const t = key => key

      expect(shallow(<DumbBillChip bill={bill} t={t} />).html()).toBe(null)
    })
  })

  it('should render a Chip', () => {
    const bill = {
      _id: 'fakebill',
      invoice: 'io.cozy.files:deadbeef'
    }

    const t = key => key

    const root = shallow(<DumbBillChip bill={bill} t={t} />)
    expect(root).toMatchSnapshot()
  })
})
