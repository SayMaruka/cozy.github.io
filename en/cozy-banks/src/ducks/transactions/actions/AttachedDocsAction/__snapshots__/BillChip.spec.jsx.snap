// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillChip should render a Chip 1`] = `
<Component
  fileId="deadbeef"
  key="deadbeef"
>
  <Chip
    component="button"
    size="small"
    theme="normal"
    variant="outlined"
  >
    <FileIcon
      className="u-flex-shrink-0"
    />
    Transactions.actions.attachedDocs.billWithoutVendor
  </Chip>
</Component>
`;
